app/components/specific/elevationProfile/ElevationProfile.js

```diff
@@ -53,16 +71,81 @@ export default class ElevationProfile extends Component {
     );
   }
 
-  // TODO: this could be done better! faster!
-  static computeDistance(edges, currentEdge) {
-    const index = edges.indexOf(currentEdge);
-    const pathDone = edges.slice(0, index);
-    return positionHelper.computeDistance(...pathDone);
-  }
+  // TODO: to be moved
+  static convertPercentToGrade = percent => {
+    if (Math.abs(percent) < 5) {
+      return ELEVATION_GRADE.SMALL;
+    } else if (Math.abs(percent) >= 5 && Math.abs(percent) < 7) {
+      return ELEVATION_GRADE.EASY;
+    } else if (Math.abs(percent) >= 7 && Math.abs(percent) < 10) {
+      return ELEVATION_GRADE.MEDIUM;
+    } else if (Math.abs(percent) >= 10 && Math.abs(percent) < 15) {
+      return ELEVATION_GRADE.DIFFICULT;
+    } else if (Math.abs(percent) >= 15) {
+      return ELEVATION_GRADE.HARD;
+    }
+    return ELEVATION_GRADE.UNKNOWN;
+  };
+
+  // TODO: to be moved
+  static convertGradeToColor = grade => {
+    if (grade === ELEVATION_GRADE.SMALL) {
+      return ELEVATION_COLORS.SMALL;
+    } else if (grade === ELEVATION_GRADE.EASY) {
+      return ELEVATION_COLORS.EASY;
+    } else if (grade === ELEVATION_GRADE.MEDIUM) {
+      return ELEVATION_COLORS.MEDIUM;
+    } else if (grade === ELEVATION_GRADE.HARD) {
+      return ELEVATION_COLORS.HARD;
+    } else if (grade === ELEVATION_GRADE.DIFFICULT) {
+      return ELEVATION_COLORS.DIFFICULT;
+    }
+    return ELEVATION_COLORS.UNKNOWN;
+  };
+
+  // TODO: to be moved
+  static addData = (...edges) => {
+    let distanceDone = 0;
+    return edges.reduce((accu, edge, index) => {
+      const length = positionHelper.computeDistance(edge);
+      distanceDone += length;
+      const elevation = (edge.dest.altitude - edge.src.altitude) / 1000;
+      const percent = elevation / length * 100;
+      const grade = ElevationProfile.convertPercentToGrade(percent);
+      const color = ElevationProfile.convertGradeToColor(grade);
+
+      const enhancedEdge = {
+        edge,
+        length,
+        percent,
+        index,
+        distanceDone,
+        grade,
+        color
+      };
+      return [...accu, enhancedEdge];
+    }, []);
+  };
+
+  // TODO: to be moved
+  static groupBy = (...edges) => {
```

Comment:
groupBy should be prototypeOn array and take a grouping fct that return the key of an item

Reply:
👍